formSchemas.js

```diff
@@ -50,6 +80,106 @@ const sinSchema = {
   },
 }
 
+const lastDayInMonth = (year,month) => {
+  return  new Date(year, month +1, 0).getDate();
+}
+
+const validBirthDateLengths = {
+  errorMessage: 'errors.login.dateOfBirth.format',
+  validate: value => {
+    const birthDateSplit = value.split('/')
+    return birthDateSplit[0].length === 4 && birthDateSplit[1].length === 2 && birthDateSplit[2].length === 2
+  }
+}
+
+const validBirthDateChars = {
+  errorMessage: 'errors.login.dateOfBirth.characters',
+  validate : value => {
+    const numAndSlash = new RegExp(/^[0-9\/]*$/)
+    return numAndSlash.test(value)
+  }
+}
+
+const validMonth = {
+  errorMessage: 'errors.login.dateOfBirth.validMonth',
+  validate : value => {
+    //month is not less than 1 or greater than 12
+    const month = parseInt(value.split('/')[1], 10);
+    return month >= 1 && month <= 12 
+  }
+}
+
+const validDay = {
+  errorMessage: 'errors.login.dateOfBirth.validDay',
+  validate : value => {
+    //day is within the acceptable range
+    const dateSplit = value.split('/')
+    //subtract one because Date for months starts at a 0 index for Jan
+    const month = parseInt(dateSplit[1]) - 1
+    const day = parseInt(dateSplit[2], 10)
+    const lastDay = lastDayInMonth(dateSplit[0], month)
+    return day >= 1 && day <= lastDay 
+  
+  }
+}
+
+const validYear = {
+  errorMessage: 'errors.login.dateOfBirth.validYear',
+  validate : value => {
+    //year is within the acceptable range. Person not older than 200, not younger than 1
+    //listen, if you've lived to 201, why are you even bothering with taxes?
+    const dateEntered = new Date(value)
+    const currentDate = new Date()
+    const year = currentDate.getFullYear()
+    const month = currentDate.getMonth()
+    const day = currentDate.getDate()
+
+    const lastYear = new Date(year - 1, month, day)
+
+    const longAgo = new Date(year -200, month, day)
+
+    return dateEntered <= lastYear && dateEntered >= longAgo
+  }
+}
+
+const isMatchingDoB = {
+  errorMessage: 'errors.login.dateOfBirth.match',
+  validate: (value, req) => {
+    /* If there is no session, always return true */
+    if (!req.session || !req.session.personal) {
+      return true
+    }
+
+    return value === req.session.personal.dob
+  }
+}
+
+const birthSchema = {
+  dateOfBirth: {
+    //pro-tip: the order of the sanitizers does matter
+    customSanitizer: {
+      options: value => {
+        //We want to remove any spaces, dash or underscores
+        return value ? value.replace(/[ \-_]*/g, '') : value
+      },
+    },
+    ...validationArray([
+      validBirthDateLengths,
+      validBirthDateChars,
+      validMonth,
+      validDay,
+      validYear,
+      isMatchingDoB,
+    ]),
+    isLength: {
+      errorMessage: 'errors.login.dateOfBirth',
+      //length with slashes
+      options: { min: 10, max: 10 },
```

Comment:
This means that people have to do `01` for days less than 10, right?

Like `1990/10/02`?

Reply:
(yes, just saw the test)

Reply:
we could change this, and quickly sanitize the value to see if it needs a `0` beforehand, if the user hasn't entered it